staking-pool-shares/src/lib.rs

```diff
@@ -0,0 +1,759 @@
+    #[init]
+    pub fn new(
+        owner_id: AccountId,
+        stake_public_key: Base58PublicKey,
+        reward_fee_fraction: RewardFeeFraction,
+    ) -> Self {
+        assert!(!env::state_exists(), "Already initialized");
+        reward_fee_fraction.assert_valid();
+        assert!(
+            env::is_valid_account_id(owner_id.as_bytes()),
+            "The owner account ID is invalid"
+        );
+        Self {
+            owner_id,
+            stake_public_key: stake_public_key.into(),
+            last_epoch_height: env::epoch_height(),
+            last_locked_account_balance: 0,
+            last_account_balance: env::account_balance(),
+            total_staked_balance: 0,
+            total_stake_shares: 0,
+            reward_fee_fraction,
+            accounts: Map::new(b"u".to_vec()),
+        }
+    }
+
+    /// Call to distribute rewards after the new epoch. It's automatically called before every
+    /// action.
+    pub fn ping(&mut self) {
+        // Checking if we need there are rewards to distribute.
+        let epoch_height = env::epoch_height();
+        if self.last_epoch_height == epoch_height {
+            return;
+        }
+        self.last_epoch_height = epoch_height;
+
+        // New total amount (both locked and unlocked balances).
+        // NOTE: We need to subtract `attached_deposit` in case `ping` called from `deposit` call.
+        let total_balance =
+            env::account_locked_balance() + env::account_balance() - env::attached_deposit();
+        // Old total amount.
+        let last_total_balance = self.last_account_balance + self.last_locked_account_balance;
+
+        // TODO: Verify if the reward can become negative.
+        let total_reward = total_balance.saturating_sub(last_total_balance);
+        if total_reward > 0 {
+            // The validation fee that the contract owner takes. If no one else is actively staking,
+            // the owner takes the full reward. This may happen due to gas rebate or when everyone
+            // else has unstaked.
+            let owners_fee = if self.total_stake_shares > 0 {
+                self.reward_fee_fraction.multiply(total_reward)
+            } else {
+                total_reward
+            };
+
+            // Distributing the remaining reward to the delegators first.
+            let remaining_reward = total_reward - owners_fee;
+            self.total_staked_balance += remaining_reward;
+
+            // Now buying "stake" shares for the contract owner at the new shares price.
+            let num_shares = self.num_shares_from_amount_rounded_up(owners_fee);
+            if num_shares > 0 {
+                // Updating owner's inner account
+                let owner_id = self.owner_id.clone();
+                let mut account = self.get_account(&owner_id);
+                account.stake_shares += num_shares;
+                self.save_account(&owner_id, &account);
```

Comment:
when do other accounts get updated?

Reply:
They get rewards from price increase

Reply:
So actually what is the point of calculating `remaining_reward`? It seems that it is not used.